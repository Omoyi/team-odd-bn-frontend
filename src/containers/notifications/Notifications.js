import React, { Component } from 'react';
import { connect } from 'react-redux';
import Moment from 'react-moment';
import '../../assets/css/notification.scss';
import { Link } from 'react-router-dom';
import Dashboard from '../../views/Dashboard/sidebar';
import picture from '../../assets/images/default_user_16px.png';
import getAllNotifications, { updateSpinnerStatus } from '../../redux/actions/notifications/getAllNotificationActions';
import verifyToken from '../../helpers/verifyToken';
import markOneAsRead from '../../redux/actions/notifications/markOneAsReadActions';

export class Notifications extends Component {
  constructor(props) {
    super(props);
    this.state = {
    };

    this.handleMarkAsRead = this.handleMarkAsRead.bind(this);
  }

  async componentDidMount() {
    const { getAllNotifications, markOneAsRead } = this.props;
    await getAllNotifications();
    const { allNotifications } = this.props;
  }

  async handleMarkAsRead(notificationId) {
    const { markOneAsRead } = this.props;
    await markOneAsRead(notificationId);
  }

  render() {
    const { allNotifications } = this.props;

    return (
      <>
        <Dashboard>
          <div className="notification-container">
            <h2>Notifications</h2>
            <hr className="title-underline" />
            <div className="all-notification-content">
              {allNotifications && allNotifications !== 'no new notification' && allNotifications.length > 0 ? (
                <>
                  {allNotifications.filter((item) => item.userId === verifyToken(localStorage.getItem('token')).id).map((item, index) => (
                    <div className="single-notification-content">
                      <div className="single-notification">
                        <div className="notification-element" key={item.id} role="button" tabIndex="0">
                          <img src={picture} alt="pic" />
                          {item.tripRequestId && (
                          <Link to={`/requests/${item.tripRequestId}`} onClick={() => this.handleMarkAsRead(item.id)}>
                            <p>{item.message}</p>
                          </Link>
                          )}
                          {item.bookingId && (
                          <Link to={`/bookings/${item.bookingId}`} onClick={() => this.handleMarkAsRead(item.id)}>
                            <p>{item.message}</p>
                          </Link>
                          )}
                        </div>
                        {!item.markRead && (
                        <div className="read-icon">
                          <span className="unread" />
                        </div>
                        )}
                      </div>
                      <div className="single-notification-date">
                        <p><Moment fromNow>{item.createdAt}</Moment></p>
                      </div>
                      <hr className="content-underline" />
                    </div>
                  ))}
                </>
              ) : <p>No Notifications found</p>}
            </div>
          </div>
        </Dashboard>
      </>
    );
  }
}

export const mapStateToProps = (state) => ({
  allNotifications: state.notifications.notifications.allNotifications,
  allNotificationsError: state.notifications.notifications.allNotificationsError,
  spinner: state.notifications.spinner,
});

export default connect(mapStateToProps, { getAllNotifications, updateSpinnerStatus, markOneAsRead })(Notifications);
