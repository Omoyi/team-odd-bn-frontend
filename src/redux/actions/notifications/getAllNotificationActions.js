import apiCall from '../../../helpers/apiCall';
import { SPINNER_STATUS, GET_ALL_NOTIFICATIONS_ERROR, GET_ALL_NOTIFICATIONS_SUCCESS } from '../../actionTypes/notificationActionTypes';

export const updateSpinnerStatus = (value) => ({
  type: SPINNER_STATUS,
  spinner: value,
});

const getAllNotifications = () => async (dispatch) => {
  const API_URL = '/users/notification';
  const HEADERS_REQUEST = {
    token: localStorage.getItem('token'),
  };

  try {
    const getAllNotificationsData = await apiCall.get(API_URL, { headers: HEADERS_REQUEST });

    dispatch(updateSpinnerStatus(false));
    return dispatch({
      type: GET_ALL_NOTIFICATIONS_SUCCESS,
      payload: getAllNotificationsData.data.data,
    });
  } catch (error) {
    dispatch(updateSpinnerStatus(false));
    return dispatch({
      type: GET_ALL_NOTIFICATIONS_ERROR,
      payload: error.response.data,
    });
  }
};

export default getAllNotifications;
