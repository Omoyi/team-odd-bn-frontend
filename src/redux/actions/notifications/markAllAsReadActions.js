import apiCall from '../../../helpers/apiCall';
import { MARK_ALL_AS_READ_ERROR, MARK_ALL_AS_READ_SUCCESS } from '../../actionTypes/notificationActionTypes';

const markAllAsRead = (notificationIds) => async (dispatch) => {
  const API_URL = '/notifications/markRead';
  const HEADERS_REQUEST = {
    token: localStorage.getItem('token'),
  };

  try {
    const markAllAsReadData = await apiCall.patch(API_URL, { notificationIds }, { headers: HEADERS_REQUEST });
    return dispatch({
      type: MARK_ALL_AS_READ_SUCCESS,
      payload: markAllAsReadData.data,
    });
  } catch (error) {
    return dispatch({
      type: MARK_ALL_AS_READ_ERROR,
      payload: error.response.data,
    });
  }
};

export default markAllAsRead;
