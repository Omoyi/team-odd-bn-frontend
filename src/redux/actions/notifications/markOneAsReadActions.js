import apiCall from '../../../helpers/apiCall';
import { MARK_ONE_AS_READ_ERROR, MARK_ONE_AS_READ_SUCCESS } from '../../actionTypes/notificationActionTypes';

const markOneAsRead = (notificationId) => async (dispatch) => {
  const API_URL = '/notifications/markRead';
  const HEADERS_REQUEST = {
    token: localStorage.getItem('token'),
  };

  try {
    const markOneAsReadData = await apiCall.patch(API_URL, { notificationIds: [notificationId] }, { headers: HEADERS_REQUEST });
    return dispatch({
      type: MARK_ONE_AS_READ_SUCCESS,
      payload: markOneAsReadData.data,
    });
  } catch (error) {
    return dispatch({
      type: MARK_ONE_AS_READ_ERROR,
      payload: error.response.data,
    });
  }
};

export default markOneAsRead;
