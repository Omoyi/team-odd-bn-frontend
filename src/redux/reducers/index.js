import { combineReducers } from 'redux';
import socialAuth from './socialAuthReducer';
import authReducers from './authReducers';
import profileReducers from './ProfileReducers';
import updateProfileReducer from './updateProfileReducer';
import getAllUsersReducer from './getAllUsersReducers';
import signoutReducer from './signoutReducer';
import tripReducers from './tripReducers';
import getAllCitiesReducers from './getAllCitiesReducers';
import accommodationReducer from './accommodation';
import multicityReducer from './multicityReducer';
import adminReducer from './admin/adminReducer';
import viewAccomodationReducer from './accommodation/viewAccomodationReducer';
import PaginationReducer from './PaginationReducer';
import searchReducer from './searchReducers/searchReducers';

import bookingReducer from './bookings';

export default combineReducers({
  viewProfile: profileReducers,
  updateProfile: updateProfileReducer,
  allUsers: getAllUsersReducer,
  socialAuth,
  auth: authReducers,
  signout: signoutReducer,
  trips: tripReducers,
  allCities: getAllCitiesReducers,
  accommodation: accommodationReducer,
  multicityTrip: multicityReducer,
  admin: adminReducer,
  allAccomodation: viewAccomodationReducer,
  pagination: PaginationReducer,
  bookings: bookingReducer,
  search: searchReducer,
});
