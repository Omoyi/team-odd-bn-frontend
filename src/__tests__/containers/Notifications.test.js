import React from 'react';
import { shallow, mount } from 'enzyme';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import dotenv from 'dotenv';
import { Notifications, mapStateToProps } from '../../containers/notifications/Notifications';
import rootReducer from '../../redux/reducers/index';
import mockData from '../../__mocks__/fileMock';
import 'jest-localstorage-mock';

dotenv.config();

const middlewares = [thunk];
let store;

const testStore = (state) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, state);
};

const setUp = (initialState = {}) => {
  store = testStore(initialState);
  const wrapper = shallow(
    <Notifications {...mockData.notificationProps} store={store} />,
  );
  return wrapper;
};

describe('Notification Test Suite', () => {
  it('Should Pass New props Successfully', () => {
    const component = setUp(mockData.notificationsMainState);
    component.setProps({ notifications: mockData.notificationProps.notifications });
  });

  it('Should return initial data', () => {
    const stores = testStore(mockData.notificationsMainState);
    localStorage.setItem('token', process.env.TESTING_TOKEN);
    expect(shallow(
      <Notifications spinner={null} store={stores} />,
    )).toMatchSnapshot();
  });

  it('Should return the initial State', () => {
    const initialState = {
      notifications: {
        notifications: {
          allNotifications: null,
          allNotificationError: null,
        },
        newNotification: null,
      },
      markOneAsRead: {
        markOne: null,
        markOneError: null,
      },
      markAsRead: {
        markAll: null,
        markAllError: null,
      },
    };
    expect(mapStateToProps(initialState).allNotifications).toEqual(null);
  });
});

describe('Should simulate mark as read', () => {
  const component = shallow(<Notifications {...mockData.notificationProps} />);
  // localStorage.setItem('token', process.env.TESTING_TOKEN);
  // component.find('.links').simulate('click');
  it('Should simulate mark as read', () => {
    const handleMarkAsRead = jest.spyOn(component.instance(), 'handleMarkAsRead');
    component.instance().handleMarkAsRead(0);
    expect(handleMarkAsRead).toHaveBeenCalledWith(0);
  });
});
