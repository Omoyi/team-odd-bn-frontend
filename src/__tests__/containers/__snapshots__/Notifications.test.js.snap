// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Test Suite Should return initial data 1`] = `ShallowWrapper {}`;
